src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/UnusedTrackerWithR8.groovy

```diff
@@ -0,0 +1,265 @@
+            private boolean expectingSeparator
+
+            @Override
+            void accept(String s, DiagnosticsHandler handler) {
+                if (classString == null && lastString == LINE_SEPARATOR) {
+                    classString        = s
+                    expectingSeparator = true
+                } else if (expectingSeparator && s == LINE_SEPARATOR) {
+                    removedClasses.add(classString)
+                    classString        = null
+                    expectingSeparator = false
+                } else {
+                    classString        = null
+                    expectingSeparator = false
+                }
+
+                lastString = s
+            }
+        })
+
+        List<String> proguardConfig = getKeepRules()
+        proguardConfig.add("-dontoptimize")
+        proguardConfig.add("-dontobfuscate")
+        proguardConfig.add("-ignorewarnings")
+
+        builder.addProguardConfiguration(proguardConfig, Origin.unknown())
+
+        builder.setProgramConsumer(new ClassFileConsumer() {
+            @Override
+            void accept(ByteDataView byteDataView, String s, DiagnosticsHandler diagnosticsHandler) {
+                String name = typeNameToExternalClassName(s)
+
+                // any class that is actually going to be written to the output
+                // must not be present in the set of removed classes.
+                // Should not really be needed but we prefer to be paranoid.
+                removedClasses.remove(name)
+
+                Path classFile = Paths.get(tmpDir.toString(), externalClassNameToInternal(name) + ".class")
+                Files.createDirectories(classFile.getParent())
+                Files.write(classFile, byteDataView.getBuffer())
+            }
+
+            @Override
+            void finished(DiagnosticsHandler diagnosticsHandler) {}
+        })
+
+        R8.run(builder.build())
+        return removedClasses
+    }
+
+    @CompileStatic
+    static String typeNameToExternalClassName(String typeName) {
+        String className = typeName.startsWith("L") && typeName.endsWith(";") ?
+            typeName.substring(1, typeName.length() - 1) :
+            typeName
+
+        return className.replaceAll("/", ".")
+    }
+
+    @CompileStatic
+    static String externalClassNameToInternal(String className) {
+        return className.replaceAll("\\.", "/")
+    }
+
+    @CompileStatic
+    List<String> getKeepRules() {
+        R8Command.Builder builder = R8Command.builder()
+
+        populateBuilderWithProjectFiles(builder)
+
+        // add all dependencies as library jars to avoid warnings
+        for (File dep : dependencies) {
+            Path path = Paths.get(dep.getAbsolutePath())
+            builder.addLibraryFiles(path)
+        }
+
+        // disable everything, we just want to get a list of
+        // all project classes.
+        List<String> configs = new ArrayList<>()
+        configs.add("-dontshrink")
```

Comment:
With `-dontshrink` isn't this just going to generate the list of the classes in the jar to be shadowed?

Reply:
yes thats the idea. I did not want to explode any jar file that is declared as input and collect classes, but rather feed R8 with the input and collect the classnames from the output. This is not necessary and could be improved.